docs/rfcs/040-Endpoint-Persistent-Unlogged-Files-Storage.md

```diff
@@ -0,0 +1,288 @@
+  Endpoints that host the data.
+- This storage system does not need to store _all_ user files, only 'known'
+  user files.
+- This storage system does not need to be hosted fully inside Computes.  
+  _Instead, this will be a separate component similar to Pageserver,
+  SafeKeeper, the S3 proxy used for dynamically loaded extensions, etc._
+
+## Impacted components (e.g. pageserver, safekeeper, console, etc)
+- Compute needs new code to load and store these files in its lifetime.
+- Console or Control Plane needs to consider this new storage system when
+  signalling the deletion of an Endpoint, Timeline, or Tenant.
+
+A new service is created: the Endpoint Persistent Unlogged Files Storage
+service.  This could be integrated in e.g. Pageserver or Control Plane, or a
+separately hosted service.
+
+## Proposed implementation
+Endpoint-related data files are managed by a newly designed service (which
+optionally is integrated in an existing service like Pageserver or Control
+Plane), which stores data directly into S3 or any blob storage of choice.
+
+Upon deletion of the Endpoint, or reassignment of the endpoint to a different
+branch, this ephemeral data is dropped: the data stored may not match the
+state of the branch's data after reassignment, and on endpoint deletion the
+data won't have any use to the user.
+
+Compute gets credentials which it can use to authenticate to this new service
+and retrieve and store data associated with that endpoint.
+
+The path of this endpoint data in S3 is initially as follows:
+
+    s3://<regional-epufs-bucket>/
+        tenant-<hex-tenant-id>/
+            tl-<hex-timeline-id>/
+                <endpoint-id>/
+                    pgdata/<file_path_in_pgdatadir>`
+
+For other blob storages an equivalent or similar path can be constructed.
+
+### Reliability, failure modes and corner cases (if relevant)
+Reliability is important, but not critical to the workings of Neon.  The data
+stored in this service will, when lost, reduce performance, but won't be a
+cause of permanent data loss - only operational metadata is stored.
+
+Most, if not all, blob storage services have sufficiently high persistence
+guarantees to cater our need for persistence and uptime. The only concern with
+blob storages is that the access latency is generally higher than local disk,
+but for the object types stored (stats files, cache state, ...) I don't think 
+this will be much of an issue.
+
+### Interaction/Sequence diagram (if relevant)
+
+In these diagrams you can replace S3 with any persistent storage device of
+choice, but S3 is chosen as well-known name of AWS' blob storage - Azure Blob
+Storage should work too, but is a much longer name.
+
+Write data:
+```mermaid
+sequenceDiagram
+    autonumber
+    participant co as Compute
+    participant ep as EPUFS
+    participant s3 as Blob Storage
+
+    co-->ep: Connect with credentials
+    co->>+ep: Store Unlogged Persistent File
```

Comment:
Let's provide a brief API spec for this EPUFS service, i.e. what are the methods and parameters we are going to have:
- PUT: tenant, timeline, endpoint, relative path, data -> json response
- GET: tenant, timeline, endpoint, relative path -> file content response
- DELETE: tenant [ timeline [ endpoint ] ] -> json response